### ISSUE TITLE
Docker Compose for Node not properly finishing the spinup

### ISSUE THREAD

**Issue by aixellent** [testcontainers/testcontainers-node#133]
I tried to spin up my Testsetup in my globalSetup file (JEST) with the following docker-compose.yml:
```
version: '3.7'

services:
  rabbitmq:
    image: rabbitmq:3-management
    container_name: rabbitmq
    environment:
      RABBITMQ_DEFAULT_USER: test
      RABBITMQ_DEFAULT_PASS: test
    volumes:
      - './rabbitmq/etc/:/etc/rabbitmq/'
      - './rabbitmq/etc/definitions.json:/etc/rabbitmq/definitions.json:ro'
    ports:
      - 5672
      - 15672
      - 25672
  api:
    image: gcr.io/gke-utility/mocks
    container_name: api
    ports:
      - 5000
    command: api/v1.yaml
```

This is just to ensure my dependencies for the service i want to Test.

Global Setup file
```
import { Duration, TemporalUnit } from 'node-duration'
import { DockerComposeEnvironment, StartedTestContainer } from 'testcontainers'
import path = require('path')

export default async (): Promise<void> => {
  const composeFilePath: string = path.basename('../.docker')
  const composeFile = 'docker-compose.yml'
  const environment = await new DockerComposeEnvironment(composeFilePath, composeFile)
    .withStartupTimeout(new Duration(200, TemporalUnit.SECONDS))
    .up()

  const rabbitmq: StartedTestContainer = environment.getContainer('rabbitmq')
  const bully: StartedTestContainer = environment.getContainer('bully')

  // THE REST OF THE SETUP

}
```

The following problem occurs after the timeout:

`
$ NODE_ENV=test jest --config=__integration-tests__/jest.config.js --testRunner='jest-circus/runner' --ci --verbose --maxWorkers=7
Error: Port 5671 not bound after 200000ms
    at /PATH_TO_PROJECT/node_modules/testcontainers/dist/wait-strategy.js:61:23
    at IntervalRetryStrategy.<anonymous> /PATH_TO_PROJECT/node_modules/testcontainers/dist/retry-strategy.js:37:28)
    at Generator.next (<anonymous>)
    at fulfilled (/PATH_TO_PROJECT/testcontainers/dist/retry-strategy.js:5:58)
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
`

The problem occurs on binding the service Ports, for RabbitMQ, additionally 

Project structure of the IT:
![image](https://user-images.githubusercontent.com/11715453/93633381-303c9c80-f9ef-11ea-9a5d-f041b1e906f3.png)

The two containers are started properly (see container Overview here)
 
![image](https://user-images.githubusercontent.com/11715453/93635136-1781b600-f9f2-11ea-941f-18cd269fcc40.png)

What can i do to get to get this running properly, i would love to use this for the service dependency configuration, especially because the rabbitMQ Preconfiguration over definitions.json would be awesome.

**Comment by cristianrgreco** [comment]
Hi @aixellent, could you confirm what version of testcontainers you're using?

**Comment by aixellent** [comment]
Hi @cristianrgreco , sorry I missed you lift this up, I am using 3.4.4.

**Comment by cristianrgreco** [comment]
I think the problem is that rabbitmq doesn't bind to all the ports it exposes. The default wait strategy of testcontainers is to wait until all of the container's exports are bound. Could you try the log wait strategy instead and let me know if it fixes this issue?

```javascript
const environment = await new DockerComposeEnvironment(composeFilePath, composeFile)
  .withWaitStrategy("rabbitmq", Wait.forLogMessage("Server startup complete"))
  .withStartupTimeout(new Duration(200, TemporalUnit.SECONDS))
  .up()
```

**Comment by aixellent** [comment]
I will try this asap and update you here, but the interesting part is that all 3 ports are available for use if I call them manually over my tools.

**Comment by cristianrgreco** [comment]
@aixellent Sure let me know 🙂
Running rabbit locally I see it exposes 5 ports, I'd need to double check but I assume by default it doesn't bind all of them

**Comment by aixellent** [comment]
@cristianrgreco 
Thanks a lot, I tried your approach and it worked for my dependency Tools (API / rabbitMQ) right now, but wouldn´t it good if we can have a similar behaviour like when we use it for DEV reasons on the local machine? My problem now is that the Service container does not work with this WaitMechanism.
I know that since version 3 of compose it is not that easy to use a manually defined startup order: https://docs.docker.com/compose/startup-order/ so that containers really wait for other containers. 
I tried the Wait for Log strategy for our service container too, but this does not work as it starts the container parallel to the rabbit what causes a error in the service startup.  

`
yarn run v1.22.5
$ NODE_ENV=test jest --config=__integration-tests__/jest.config.js --testRunner='jest-circus/runner' --ci --verbose --maxWorkers=7
Error: (HTTP code 409) unexpected - Container 8169527d30c69904fbfc5c93db08cd122eac27b878890a333bdbbfe3e0a279f7 is not running
    at /THE_SERVICE/node_modules/docker-modem/lib/modem.js:301:17
    at getCause (/THE_SERVICE/node_modules/docker-modem/lib/modem.js:331:7)
    at Modem.buildPayload (/THE_SERVICE/node_modules/docker-modem/lib/modem.js:300:5)
    at IncomingMessage.<anonymous> (/THE_SERVICE/node_modules/docker-modem/lib/modem.js:275:14)
    at IncomingMessage.emit (events.js:208:15)
    at endReadableNT (_stream_readable.js:1168:12)
    at processTicksAndRejections (internal/process/task_queues.js:77:11)
error Command failed with exit code 1.
`

What means that the container breaks the UP of the testcontainers flow. Is there anything i can do here? The container just includes a small node application which on startup connects to the RMQ.

If this is not possible to pipe it properly i can live with it and create the service container from dockerfile or manually by generic container structure.

**Comment by cristianrgreco** [comment]
Hi @aixellent, I think there are a couple options here:

1. Change the service so that if the connection fails it retries (recommended by Docker https://docs.docker.com/compose/startup-order/):

> The problem of waiting for a database (for example) to be ready is really just a subset of a much larger problem of distributed systems. In production, your database could become unavailable or move hosts at any time. Your application needs to be resilient to these types of failures.
>
> To handle this, design your application to attempt to re-establish a connection to the database after a failure. If the application retries the connection, it can eventually connect to the database.
>
> The best solution is to perform this check in your application code, both at startup and whenever a connection is lost for any reason.

2. For a `GenericContainer`, we're able to use a `HealthCheck` wait strategy as follows:

```typescript
.withHealthCheck({
  test: "curl -f http://localhost || exit 1",
  interval: new Duration(1, TemporalUnit.SECONDS),
  timeout: new Duration(3, TemporalUnit.SECONDS),
  retries: 5,
  startPeriod: new Duration(1, TemporalUnit.SECONDS)
})
.withWaitStrategy(Wait.forHealthCheck())
.start();
```

This isn't supported yet for docker-compose but would something like this solve the problem if you were able to use it as the wait strategy for the service container? I'm thinking the health check would be the service trying to connect to rabbit and it could try a number of times over a time period

**Comment by aixellent** [comment]
@cristianrgreco 
Thanks for this advice, i tried the first option, but if it fails once, it breaks the cycle, the only difference is the follwoing

```
yarn run v1.22.5
$ NODE_ENV=test jest --config=__integration-tests__/jest.config.js --testRunner='jest-circus/runner' --ci --verbose --maxWorkers=7
Error: (HTTP code 409) unexpected - Container 8169527d30c69904fbfc5c93db08cd122eac27b878890a333bdbbfe3e0a279f7 **is restarting**
    at /THE_SERVICE/node_modules/docker-modem/lib/modem.js:301:17
    at getCause (/THE_SERVICE/node_modules/docker-modem/lib/modem.js:331:7)
    at Modem.buildPayload (/THE_SERVICE/node_modules/docker-modem/lib/modem.js:300:5)
    at IncomingMessage.<anonymous> (/THE_SERVICE/node_modules/docker-modem/lib/modem.js:275:14)
    at IncomingMessage.emit (events.js:208:15)
    at endReadableNT (_stream_readable.js:1168:12)
    at processTicksAndRejections (internal/process/task_queues.js:77:11)
error Command failed with exit code 1.
```

But no problem with that, i got it working with 2 COMPOSE Files, one for the dependencies and one for the service, they are bound over a network together so they can interact without problems. 

But NOW i have the following problem on TEARDOWN

```
 PASS  __integration-tests__/__tests__/health.test.ts
  Integration tests
    ✓ TC 0: Health Endpoint check (35ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        0.978s, estimated 2s
Ran all test suites.
jest-html-reporter >> Report generated (/Users/schmitz/Development/Projects/shop-apotheken-offer-adapter/test-report.html)

All tests passed

__tests__/health.test.ts
  Integration tests
    ✓ TC 0: Health Endpoint check

{"level":"warn","message":"🧨 Tear down Integration Test Environment 🧨","module_name":"the-service","module_version":"0.0.1","timestamp":"2020-09-21T10:07:11.688Z"}
Stopping rabbitmq ... done
Stopping api    ... done
Found orphan containers (the-service) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.
Removing rabbitmq ... done
Removing api    ... done
Removing network __mocks___default
error while removing network: network __mocks___default id c378a1b6d8ef386a99beb2f6e439b246299c92504b1596f680170df50e704f34 has active endpoints
    at StartedDockerComposeEnvironment.<anonymous> (/path/Development/Projects/the-service/node_modules/testcontainers/dist/docker-compose-environment.js:158:23)
    at Generator.throw (<anonymous>)
    at rejected (/path/Development/Projects/the-service/node_modules/testcontainers/dist/docker-compose-environment.js:25:65)
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Is there a possibility to kill the network o.e. a container which dows not stop properly (as this is just for a testcycle)

**Comment by cristianrgreco** [comment]
@aixellent Apologies just a few questions so I'm sure I understand correctly:

> i tried the first option, but if it fails once, it breaks the cycle

What's meant by it breaks the cycle? My understanding is:
- We have a rabbit container which takes a while to start up
- We have a service container that tries to connect to rabbit immediately. Rabbit isn't ready at this point which causes this container to fail to start.

If this is the case, then I was suggesting that the service container continually retries to connect to rabbit until it is ready. If it does this then as long as rabbit comes up successfully, both containers will eventually start successfully.

If this change isn't possible then do you think the health check wait strategy I proposed in the previous comment would solve the problem?

**Comment by aixellent** [comment]
@cristianrgreco 
Sorry for the confusion, my bad.
I tried the approach here: https://github.com/testcontainers/testcontainers-node/issues/133#issuecomment-695332527 which is working fine for the dependencies (rabbit & API) now.
Then i thought about how to fix the problem with the SERVICE Container which needs to either wait for the rabbit to be ready OR to reconstruct the startup mechanism of it for facing this problem. 
Waiting is not possible easily with docker compose as you said, so this option has gone. The second approach is not feasible either, why? Because our K8s Strategy says that we should BREAK the service on errors so it can be restarted by the orchestrator, so the service design could not be adapted here only for testreasons (maybe it could by ENVs but we internally agreed on that this is quite bad smell stuff). So i configured the service in the compose file with `restart:always` what did the restart part you described, BUT initially it fails and breaks the spinup and delivers the following message:

```
yarn run v1.22.5
$ NODE_ENV=test jest --config=__integration-tests__/jest.config.js --testRunner='jest-circus/runner' --ci --verbose --maxWorkers=7
Error: (HTTP code 409) unexpected - Container 8169527d30c69904fbfc5c93db08cd122eac27b878890a333bdbbfe3e0a279f7 **is restarting**
```     

So i came to the point that i outsourced the SERVICE to a second compose file which is started sequentially after RABBIT and API have been spinned up successfully (with the help of your WaitForLog proposal)

![image](https://user-images.githubusercontent.com/11715453/93784136-e5a86380-fc2c-11ea-9798-c1be16324379.png)


That is working quite fine, as the services are bound over a network i only have one problem left, on `globalTeardown` the service container / service container compose environment cannot be stopped / downed because of the network which is automatically created with the name `__mocks___default` refuses this as it says that there are active connections left.

Error:
```
 PASS  __integration-tests__/__tests__/health.test.ts
  Integration tests
    ✓ TC 0: Health Endpoint check (35ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        0.978s, estimated 2s
Ran all test suites.
jest-html-reporter >> Report generated (/pathProjects/the-service/test-report.html)

All tests passed

__tests__/health.test.ts
  Integration tests
    ✓ TC 0: Health Endpoint check

{"level":"warn","message":"🧨 Tear down Integration Test Environment 🧨","module_name":"the-service","module_version":"0.0.1","timestamp":"2020-09-21T10:07:11.688Z"}
Stopping rabbitmq ... done
Stopping api    ... done
Found orphan containers (the-service) for this project. If you removed or renamed this service in your compose file, you can run this command with the --remove-orphans flag to clean it up.
Removing rabbitmq ... done
Removing api    ... done
Removing network __mocks___default
error while removing network: network __mocks___default id c378a1b6d8ef386a99beb2f6e439b246299c92504b1596f680170df50e704f34 has active endpoints
    at StartedDockerComposeEnvironment.<anonymous> (/path/Development/Projects/the-service/node_modules/testcontainers/dist/docker-compose-environment.js:158:23)
    at Generator.throw (<anonymous>)
    at rejected (/path/Development/Projects/the-service/node_modules/testcontainers/dist/docker-compose-environment.js:25:65)
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

So the question from my side would be, will there be an option to pass params like `--remove-orphans` or `--rmi type` in the future? I know this looks dirty too, but in some cases this would be a great feature.

**Comment by cristianrgreco** [comment]
I ran a test locally with `--remove-orphans` but it still fails to stop docker-compose because the network is in use.

According to the docs:
- `docker-compose stop`: The docker-compose stop command will stop your containers, but it won’t remove them.
- `docker-compose down`: The docker-compose down command will stop your containers, but it also removes the stopped containers **as well as any networks that were created.**

This means that if you have two running docker-composes sharing a network, and you down one of them, docker will try to kill the network, which is in use, and will fail.

Based on this, I think what we might want to do is:
1. `docker-compose up` the dependencies
2. `docker-compose up` the service
3. `docker-compose stop` the service
4. `docker-compose down` the dependencies.

What do you think?

**Comment by aixellent** [comment]
@cristianrgreco 
That idea will fit the needs more properly, but therefore the DockerEnvironment should get the `stop` option or the down option runs stop at first for all services in the environment.
As said i actually found a solution by manually chain the TearDown in Promises so that it does not throw exceptions, but this would be really nice to have.

**Comment by cristianrgreco** [comment]
Hey @aixellent, take a look at the referenced PR - unless you see any issues with it I'll release 🙂

**Comment by aixellent** [comment]
Hi @cristianrgreco ,
Approved your PR, thanks for the support and the fast implementation.
🎉 Everything running properly now

**Comment by cristianrgreco** [comment]
@aixellent Thanks for raising the issue, this has been released in 4.1.0 🙂
